service/collector/release.go

```diff
@@ -0,0 +1,102 @@
+package collector
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/giantswarm/k8sclient/v3/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	gaugeValue float64 = 1
+)
+
+const (
+	namespace = "release_operator"
+	subsystem = "release"
+)
+
+var (
+	ReleaseDesc *prometheus.Desc = prometheus.NewDesc(
+		// TODO: What is the most descriptive name for this? Maybe change to "states" or "statuses"?
+		prometheus.BuildFQName(namespace, subsystem, "info"),
+		"Metrics for Release statuses.",
+		[]string{
+			labelName,
+			labelNamespace,
+			labelState,
+			labelReady,
+		},
+		nil,
+	)
+)
+
+type ReleaseCollector struct {
+	k8sClient k8sclient.Interface
+	logger    micrologger.Logger
+}
+
+type ReleaseCollectorConfig struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+}
+
+func NewReleaseCollector(config ReleaseCollectorConfig) (*ReleaseCollector, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	rc := &ReleaseCollector{
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+	}
+
+	return rc, nil
+}
+
+func (r *ReleaseCollector) Collect(ch chan<- prometheus.Metric) error {
+	ctx := context.Background()
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "collecting metrics")
+
+	err := r.collectReleaseStatus(ctx, ch)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "finished collecting metrics")
+	return nil
+}
+
+func (r *ReleaseCollector) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- ReleaseDesc
+	return nil
+}
+
+func (r *ReleaseCollector) collectReleaseStatus(ctx context.Context, ch chan<- prometheus.Metric) error {
+	releases, err := r.k8sClient.G8sClient().ReleaseV1alpha1().Releases().List(metav1.ListOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, release := range releases.Items {
+		ch <- prometheus.MustNewConstMetric(
```

Comment:
Is there a reason that we're using `MustNewConstMetric` instead of `NewConstMetric`? Does it make sense to log and propagate the error back instead of directly panicking here?

Reply:
Honestly not sure, it seems to be like this everywhere else we collect metrics this way but I don't know if that's intentional. It probably would be better to log and recover but let's ask @giantswarm/sig-monitoring if there's a reason to do it this way.

Reply:
hi hi hi

from https://godoc.org/github.com/prometheus/client_golang/prometheus#NewConstMetric:
```
NewConstMetric returns an error if the length of labelValues is not consistent with the variable labels in Desc or if Desc is invalid
```

On the whole, that's pretty hard to fuck up. Sure, I don't think we have any 100% rule that you must use `MustNewConstMetric` over `NewConstMetric`, but I find it only really errors if you've just written the thing wrong, as opposed to some runtime issue.

Reply:
In this way we kind of depend on that behaviour not changing, but it's probably fine in this case. Especially if we're using it everywhere already 🤷‍♀️ I was just curious